sql: name PRIMARY KEY constraints (table_name)_pkey by default

otan opened the pull request:
resolves https://github.com/cockroachdb/cockroach/issues/71087

See individual commits for details.
Split up to make reviewing "easier"; more complex changes are separate commits.

otan committed:
sql: name PRIMARY KEY constraints `(table_name)_pkey` by default

Note that legacy system tables retain their legacy PRIMARY name.

Release note (sql change): Newly created tables now have
`<table_name>_pkey` by default as their index/constraint name.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70604)
<!-- Reviewable:end -->

otan committed:
cdctest: fix broken tests after PRIMARY KEY constraint rename

Broken up for ease of reviewing; this does a little bit more than the
mechanical work.

Release note: None

otan committed:
importccl: change MySQL PRIMARY KEY index name on import

To match PostgreSQL, we should rename the PRIMARY KEY index name
when importing MySQL dumps.

Release note: None

otan committed:
tpcc: remove `@primary` index hint for check3325

Soon, we will no longer have `@primary` as the PRIMARY KEY lookup path. To
prepare for the eventuality TPC-C may be used on an older or newer
version of cockroach, remove the `@primary` index hint.

Release note: None

otan committed:
partitionccl: fix TestGenerateSubzoneSpans

This one is a doozy because it relies on using %s for the table name. As
such, hack around it by replace `@primary` with the given table name.

Release note: None

dt commented:
Nice! Just wanted to check when you were thinking of merging this? Would it be onerous to wait until we tag a 21.2 RC? We (bulk at least) are trying to avoid unnecessary refactor churn in our packages until the RC is ready, just in case we discover a last-minute severe and release-blocking issue that requires an urgent fix, so that that fix doesn't risk being delayed by unclean backports.

otan commented:
this can wait (but oh god the pain of rebasing)

postamar reviewed (approved):
This all seems fine to me, the only comment I have pertains to maintaining the legacy naming of the primary index of the system tables. 

Caveat: I haven't looked at the slew of changes in the expected test output definitions but then I doubt anyone ever will. 

As far as I'm concerned this is a welcome change, I'm just curious as to what motivates it, as I see no linked issue anywhere. Do I guess correctly that naming them all "primary" was causing problems when importing mysql dumps?

rafiss commented:
This was motivated by trying to get the https://www.prisma.io/ test suite passing against CRDB

otan commented on pkg/sql/catalog/tabledesc/structured.go:
```diff
@@ -49,8 +49,8 @@ type Mutable struct {
-	// PrimaryKeyIndexName is the name of the index for the primary key.
-	PrimaryKeyIndexName = "primary"
+	// LegacyPrimaryKeyIndexName is the pre 22.1 default PRIMARY KEY index name.
+	LegacyPrimaryKeyIndexName = "primary"
```
I'd hate to run that migration! In particular, the hardcoded descriptors and sql schema have to be kept in sync as well. That will be challenging in your example.

postamar replied:
It's not that bad, I don't think. There's no need to version-gate the system table schemas in `systemschema` as these are used for bootstrapping new clusters, changes in there are supposed to be backward compatible. In this case changing the name isn't backward compatible but if we never refer to the primary index of a system table by name anywhere then that doesn't matter and it's essentially a no-op. I suspect that we don't, ever. The migration would be in case if ever we chose to in the future.

Just to be clear I'm OK with just leaving their names be. I wanted to understand if there was a deeper motivation that I'd missed.

As for making sure not to miss tables, querying `SHOW TABLES FROM system` and iterating over the results should take care of that.

otan committed:
*: fix mechanical tests for changing PRIMARY INDEX default constraint name

Separated out from logic changes (previous commit), can squash before
merging.

Release note: None

otan commented:
bors r=postamar

RaduBerinde commented:
The rename can have important consequences: a user who is using a `@primary` index hint will see their queries fail once they recreate the schema against the new version. Users will consider this a "breaking change".

I propose that we add some code to continue to support `@primary` to refer to the primary key (as a special case). I don't see the downside to accepting it (rather than erroring out).

otan commented:
if they re-use the schema that was dumped, it will have `CONSTRAINT "primary" PRIMARY KEY ....` no?
i feel like we'd have to differentiate `@primary` and `@"primary"` otherwise as someone may name their non-PK `primary`

RaduBerinde commented:
> if they re-use the schema that was dumped, it will have CONSTRAINT "primary" PRIMARY KEY .... no?

We have seen customers recreate their schema anew using their own tools/scripts. Deploying a fresh instance of their app on a test cluster is their first step to validate that their app still works with a new version.

> i feel like we'd have to differentiate @primary and @"primary" otherwise as someone may name their non-PK primary

My proposal is to match `primary` to the primary key if it matches nothing else. Only cases that would error previously would be affected.

otan commented:
that will probs work for PKs, do we care about what happens to FKs?
cc @rafiss in case you have an opinion too

RaduBerinde commented:
FKs don't matter as much, they can't be part of queries as hints.

rafiss commented:
I'd be fine to add that fallback name resolution logic. So we aren't going to be changing anything about how it shows up in pg_catalog introspection tables, right?

Let's get it tracked in a new github issue

RaduBerinde commented:
@otan is all action, it's already done (https://github.com/cockroachdb/cockroach/pull/72534).
